Kernel/Bus/USB/xHCI/xHCIController.cpp

```diff
@@ -0,0 +1,1439 @@
+            || device_identifier.class_code() != PCI::ClassID::SerialBus
+            || device_identifier.subclass_code() != PCI::SerialBus::SubclassID::USB
+            || device_identifier.prog_if() != PCI::SerialBus::USBProgIf::EHCI)
+            return;
+        ehci_controller_found = true;
+    }));
+    if (!ehci_controller_found)
+        return;
+    dmesgln_pci(*this, "Switching Intel chipset USB ports to xHCI instead of EHCI");
+    SpinlockLocker const locker(device_identifier().operation_lock());
+    // Enable USB3 Ports
+    write32_locked(device_identifier(), intel_xhci_usb3_port_super_speed_enable_offset, read32_locked(device_identifier(), intel_xhci_usb3_port_routing_mask_offset));
+    // Enable USB2 Ports
+    write32_locked(device_identifier(), intel_xhci_usb2_port_routing_offset, read32_locked(device_identifier(), intel_xhci_usb2_port_routing_mask_offset));
+}
+
+ErrorOr<void> xHCIController::find_port_max_speeds()
+{
+    // At least one of these capability structures is required for all xHCI implementations.
+    auto* extended_capabilities_pointer = m_registers_mapping.ptr() + (m_capability_registers.capability_parameters_1.xHCI_extended_capabilities_pointer << 2);
+    if (extended_capabilities_pointer == 0)
+        return EINVAL;
+
+    ExtendedCapability volatile* extended_capability = nullptr;
+    Vector<SupportedProtocolExtendedCapability volatile*> supported_protocol_capabilities;
+    while (true) {
+        extended_capability = reinterpret_cast<ExtendedCapability volatile*>(extended_capabilities_pointer);
+        if (extended_capability->capability_id == ExtendedCapability::CapabilityID::Supported_Protocols)
+            TRY(supported_protocol_capabilities.try_append(reinterpret_cast<SupportedProtocolExtendedCapability volatile*>(extended_capability)));
+        u32 const next_extended_capability_offset = extended_capability->next_xHCI_extended_capability_pointer << 2;
+        if (next_extended_capability_offset == 0)
+            break;
+        extended_capabilities_pointer += next_extended_capability_offset;
+    };
+
+    if (supported_protocol_capabilities.is_empty())
+        return EINVAL;
+
+    for (auto* supported_protocol_capability : supported_protocol_capabilities) {
+        if (supported_protocol_capability->name_string != SupportedProtocolExtendedCapability::usb_name_string)
+            continue;
+        auto major_revision = supported_protocol_capability->major_revision;
+        if (major_revision < 0x02 || major_revision > 0x03)
+            continue;
+        auto minor_revision = supported_protocol_capability->minor_revision;
+        if (major_revision == 0x02 && minor_revision != 0x00)
+            continue;
+        if (major_revision == 0x03 && minor_revision > 0x20)
+            continue;
+        auto offset = supported_protocol_capability->compatible_port_offset;
+        if (offset < 1 || offset > m_ports)
```

Comment:
oops, switched this to use the theortical maximum instead of that value, since there's nothing to lose saving the values even if we don't actually use them.